### PR TITLE
fix: Don't emit incorrect roots in Merkle tree batch append

### PR DESCRIPTION
**Pull request by vadorovsky**
- Before this change, we were appending `current_node` at the end of each leaf append to the `roots`. This was incorrect, because during appends of non-terminal leaves, we skip computing of upper Merkle path hashes if possible - when we know that the following appends are going to overwrite these nodes.
- To fix that, for all non-terminal leaves, emit a "zero root" (`[0u8; 32]`). The most correct thing would be not emitting such roots at all, but unfortunately, that would break clients and forester, which rely on co-relation between sequence number (which is the only information that Photon returns) and root index.

Other changes:

- Remove `ChangelogEvent` - it's not used anymore. Nowadays, we emit only `PublicTransactionEvent` and it's enough for Photon.
- Modify `Changelog`:
  - Remove `root`. It's a redundant information, there is already a field for `roots` in `ConcurrentMerkleTree`.
  - Change the type of `path` from `[[u8; 32]]` to `[Option<[u8;32>]]`.
- Stop backfilling the non-terminal changelog entries, because it's confusing and requires additional operations (copying nodes across changelog paths).
  - Instead, make each path node an `Option` and when `update_proof` encounters and empty node, let it continue the loop - the subsequent changelog entries should be able to fill up the missing node.
- Implement `Debug` trait and `as_slice` method for `CyclicBoundedVec`, it makes debugging of failing tests much easier.

### CODE REVIEW

File: merkle-tree/concurrent/src/lib.rs [hunk 1]
```diff
@@ -623,23 +616,15 @@ where
...
 
                 current_index /= 2;
             }
 
-            self.changelog[changelog_index].root = current_node;
-
-            self.roots.push(current_node);
+            if leaf_i == leaves.len() - 1 {
+                self.roots.push(current_node);
+            }
```
Comment (ananas-block) [inline merkle-tree/concurrent/src/lib.rs:627]:
```suggestion
            } else {
             self.push([0u8;32]);
            }
```
We need to push the same number of roots as we append leaves.
Photon only returns the sequence number and we use it in the js client to derive the root index.
This is likely the reason why cli tests fail.

**Review by ananas-block — COMMENTED** [review]
Thanks for the fix, see comment for a potential fix for ci.

File: merkle-tree/concurrent/src/lib.rs (hunk 1 above)
Comment (vadorovsky) [reply merkle-tree/concurrent/src/lib.rs:627]: If we need to push the same number of roots, then we either need to:

- push the same root N times (feels really stupid...)
- compute the whole Merkle paths (we lose performance)

I guess you had the 1st solution in mind?

Comment (vadorovsky) [reply merkle-tree/concurrent/src/lib.rs:632]: Done

**Review by ananas-block — APPROVED** [review]

**Pull request merged**